src/ZODB/scripts/zodbcmp.py

```diff
@@ -0,0 +1,213 @@
+        attr1 = getattr(txn1, attr)
+        attr2 = getattr(txn2, attr)
+        if attr1 != attr2:
+            return 1
+
+    # data
+    objv1 = txnobjv(txn1)
+    objv2 = txnobjv(txn2)
+    if len(objv1) != len(objv2):
+        return 1
+
+    for obj1, obj2 in zip(objv1, objv2):
+        for attr in ('oid', 'data', 'data_txn'):
+            attr1 = getattr(obj1, attr)
+            attr2 = getattr(obj2, attr)
+            if attr1 != attr2:
+                return 1
+
+    return 0
+
+
+# compare two storages
+# 0 - equal, 1 - non-equal
+def storcmp(stor1, stor2, tidmin, tidmax, verbose=False):
+    iter1 = stor1.iterator(tidmin, tidmax)
+    iter2 = stor2.iterator(tidmin, tidmax)
+
+    Tprev = time()
+    txncount = 0
+    while 1:
+        txn1, ok1 = nextitem(iter1)
+        txn2, ok2 = nextitem(iter2)
+
+        # comparison finished
+        if not ok1 and not ok2:
+            if verbose:
+                print("equal")
+            return 0
+
+        # one part has entry not present in another part
+        if txn1 is None or txn2 is None or txn1.tid != txn2.tid:
+            if verbose:
+                tid1 = txn1.tid if txn1 else inf
+                tid2 = txn2.tid if txn2 else inf
+                l = [(tid1, 1,2), (tid2, 2,1)]
+                l.sort()
+                mintid, minstor, maxstor = l[0]
+                print("not-equal: tid %s present in stor%i but not in stor%i" % (
+                        ashex(mintid), minstor, maxstor))
+            return 1
+
+        # show current comparison state and speed
+        if verbose:
+            txncount += 1
+            T = time()
+            if T - Tprev > 5:
+                print("@ %s  (%.2f TPS)" % (ashex(txn1.tid), txncount / (T - Tprev)))
+                Tprev = T
+                txncount = 0
+
+        # actual txn comparison
+        tcmp = txncmp(txn1, txn2)
+        if tcmp:
+            if verbose:
+                print("not-equal: transaction %s is different")
+            return 1
+
+
+# ----------------------------------------
+import ZODB.config
+import sys, getopt
+import traceback
+
+def usage(out):
+    print("""
+Usage: zodbcmp [OPTIONS] <storage1> <storage2> [tidmin..tidmax]
+Compare two ZODB databases.
+
+<storageX> is a file with ZConfig-based storage definition, e.g.
+
+    %import neo.client
+    <NEOStorage>
+        master_nodes    ...
+        name            ...
+    </NEOStorage>
+
+Options:
+
+    -v  --verbose   increase verbosity
+    -h  --help      show this help
+""", file=out)
+
+# tidmin..tidmax -> (tidmin, tidmax)
+def parse_tidrange(tidrange):
+    tidmin, tidmax = tidrange.split("..")
+    tidmin = tidmin.decode('hex')
+    tidmax = tidmax.decode('hex')
+
+    # empty tid means -inf / +inf respectively
+    # ( which is None in IStorage.iterator() )
+    if not tidmin:
+        tidmin = None
+    if not tidmax:
+        tidmax = None
+
+    return tidmin, tidmax
```

Comment:
Thanks for the `or None` hint. However since .split("..") and .decode("hex") raise different exceptions, for proper error reporting I decided to wrap both & hex-decode actions and reaise dedicated exception so users will get e.g.

```
E: invalid tidrange: q..
```

instead of

```
Traceback (most recent call last):
  File "./zodbcmp.py", line 205, in <module>
    main()
  File "./zodbcmp.py", line 192, in main
    tidmin, tidmax = parse_tidrange(argv[2])
  File "./zodbcmp.py", line 153, in parse_tidrange
    tidmin = tidmin.decode('hex')
  File ".../lib/python2.7/encodings/hex_codec.py", line 42, in hex_decode
    output = binascii.a2b_hex(input)
TypeError: Odd-length string
```

See the change in interdiff for next patch revision.